File: src/services/github.ts
import { callTauri } from './tauri'
import type {
  GitHubQueueStatus,
  GitHubRateLimitStatus,
  GitHubRelease,
  OwnerRepositoriesResponse,
} from '../types'

const inFlightRequests = new Map<string, Promise<unknown>>()
let activeRequestCount = 0

function dedupeGithubRequest<T>(key: string, request: () => Promise<T>): Promise<T> {
  const currentRequest = inFlightRequests.get(key)
  if (currentRequest) return currentRequest as Promise<T>

  activeRequestCount += 1
  const pending = request().finally(() => {
    activeRequestCount = Math.max(0, activeRequestCount - 1)
    inFlightRequests.delete(key)
  })
  inFlightRequests.set(key, pending)
  return pending
}

export function getGithubQueueStatus(): GitHubQueueStatus {
  return {
    active: activeRequestCount,
    queued: 0,
    concurrency: activeRequestCount,
    highPriority: 0,
    normalPriority: 0,
    pausedUntil: null,
  }
}

export async function listOwnerRepositories(
  owner: string,
  page = 1,
  releasesOnly = false,
  forceRefresh = false,
): Promise<OwnerRepositoriesResponse> {
  const normalizedOwner = owner.trim().toLowerCase()
  const key = `owner:${normalizedOwner}:${page}:${releasesOnly}:${forceRefresh}`
  return dedupeGithubRequest(key, () =>
    callTauri<OwnerRepositoriesResponse>('list_owner_repositories', {
      owner,
      page,
      releasesOnly,
      forceRefresh,
    }),
  )
}

export async function getReleases(
  owner: string,
  repo: string,
  forceRefresh = false,
): Promise<GitHubRelease[]> {
  const key = `releases:${owner.trim().toLowerCase()}/${repo.trim().toLowerCase()}:${forceRefresh}`
  return dedupeGithubRequest(key, () =>
    callTauri<GitHubRelease[]>('get_releases', { owner, repo, forceRefresh }),
  )
}

export async function clearGithubCache(): Promise<void> {
  inFlightRequests.clear()
  return callTauri('clear_github_cache')
}

export async function getGithubRateLimitStatus(): Promise<GitHubRateLimitStatus> {
  return callTauri<GitHubRateLimitStatus>('get_github_rate_limit_status')
}

File: src-tauri/src/commands/github.rs
use tauri::State;

use crate::github::models::{OwnerRepositoriesResponse, Release};
use crate::github::GitHubRateLimitStatus;
use crate::AppState;

#[tauri::command]
pub async fn list_owner_repositories(
    owner: String,
    page: Option<u32>,
    releases_only: Option<bool>,
    force_refresh: Option<bool>,
    state: State<'_, AppState>,
) -> Result<OwnerRepositoriesResponse, String> {
    state
        .github_client
        .list_owner_repositories(
            &owner,
            page.unwrap_or(1),
            releases_only.unwrap_or(true),
            force_refresh.unwrap_or(false),
        )
        .await
}

#[tauri::command]
pub async fn get_releases(
    owner: String,
    repo: String,
    force_refresh: Option<bool>,
    state: State<'_, AppState>,
) -> Result<Vec<Release>, String> {
    state
        .github_client
        .get_releases(&owner, &repo, force_refresh.unwrap_or(false))
        .await
}

#[tauri::command]
pub async fn clear_github_cache(state: State<'_, AppState>) -> Result<(), String> {
    state.github_client.clear_cache();
    Ok(())
}

#[tauri::command]
pub async fn get_github_rate_limit_status(
    state: State<'_, AppState>,
) -> Result<GitHubRateLimitStatus, String> {
    Ok(state.github_client.rate_limit_status())
}
